resourceIndexer/main.js

```diff
@@ -0,0 +1,113 @@
+
+import { ArgumentParser } from 'argparse';
+import fs from 'fs';
+import { coreS3Data, stagingS3Data } from "./coreStagingS3.js";
+import {logger} from './logger.js';
+import zlib from 'zlib';
+import { promisify } from 'util';
+import { ResourceIndexerError } from './errors.js';
+
+const gzip = promisify(zlib.gzip)
+
+/**
+ * We define a number of collections which each represent some listing of
+ * nextstrain resources. The actual details are deferred to the provided
+ * collection objects - e.g. they may represent a GitHub repo listing, an S3
+ * inventory. Each of these collections provides functions which allow items
+ * (files) across collections to be collected into a master list of resources
+ * using three identifiers: source, resourceType and resourcePath. The intention
+ * is for source to parallel the information in the corresponding Source
+ * (sub-)class and resourcePath to parallel the information in the Resource
+ * (sub-)class.
+ *
+ * Currently only core & staging sources + datasets & intermediates are part of
+ * the index. For instance, the core WNV/NA (nextstrain.org/WNV/NA) dataset is
+ * indexed like so:
+ *
+ *    core → dataset → WNV/NA → versions -> [
+ *          {date: "2021-04-08", fileUrls: {main: ...},
+ *          {date: "2019-08-30", fileUrls: {meta: ..., tree: ...}
+ *    ]
+ *
+ */
+const COLLECTIONS = [
+  coreS3Data,
+  stagingS3Data,
+];
+
+function parseArgs() {
+  const argparser = new ArgumentParser({
+    description: `
+      Fetch file lists from a number of provided collections (e.g. S3 inventories) and collect them into
+      resources. Resources are organised in a hierarchical fashion via source → resourceType → resourcePath.
+      Each resource contains a list of available versions, where applicable.
+      The output JSON is intended for consumption by the nextstrain.org server.
+    `,
+  });
+  argparser.addArgument("--local", {action: 'storeTrue',
+    help: 'Access a local copy of S3 inventories within ./devData/. See docstring of fetchInventoryLocal() for expected filenames.'})
+  argparser.addArgument("--collections", {metavar: "<name>", type: "string", nargs: '+',
+    help: "Only fetch data from a subset of collections. Source names are those defined in COLLECTIONS"});
+  argparser.addArgument("--output", {metavar: "<json>", required: true})
+  argparser.addArgument("--indent", {action: 'storeTrue', help: 'Indent the output JSON'})
+  argparser.addArgument("--gzip", {action: 'storeTrue', help: 'GZip the output JSON'})
+  argparser.addArgument("--verbose", {action: 'storeTrue', help: 'Verbose logging'})
+
+  return argparser.parseArgs();
+}
+
+
+main(parseArgs())
+  .catch((err) => {
+    logger.error(err.message);
+    if (!(err instanceof ResourceIndexerError)) {
+      console.trace(err);
+    }
+  })
```

Comment:
This needs to set `process.exitCode = 1` (or similar) so that anything running the indexer knows it failed.

Reply:
Thanks - changed.